src/Repository/ProductCommentCriterionRepository.php

```diff
@@ -43,15 +59,164 @@ class ProductCommentCriterionRepository
+    {
+        $res = 0;
+
+        $criterionType = $criterion->getType();
+
+        $this->remove($criterion, true);
+
+        $res += $this->deleteLangs($criterion);
+
+        if ($criterionType == ProductCommentCriterion::CATEGORIES_TYPE) {
+            $res += $this->deleteCategories($criterion);
+        } elseif ($criterionType == ProductCommentCriterion::PRODUCTS_TYPE) {
+            $res += $this->deleteProducts($criterion);
+        }
+
+        $res += $this->deleteGrades($criterion);
+
+        return $res;
+    }
+
+    /* Update a criterion and Update its manual relation _lang, _category, _product, _grade */
+    public function update(ProductCommentCriterion $criterion): int
+    {
+        $res = 0;
+
+        $criterionType = $criterion->getType();
+
+        $this->getEntityManager()->persist($criterion);
+        $this->getEntityManager()->flush();
+
+        $res += $this->deleteLangs($criterion);
+        $res += $this->updateLangs($criterion);
+
+        if ($criterionType == ProductCommentCriterion::CATEGORIES_TYPE) {
+            $res += $this->deleteCategories($criterion);
+            $res += $this->updateCategories($criterion);
+        } elseif ($criterionType == ProductCommentCriterion::PRODUCTS_TYPE) {
+            $res += $this->deleteProducts($criterion);
+            $res += $this->updateProducts($criterion);
+        }
+
+        return $res;
+    }
+
+    private function updateLangs($criterion): int
+    {
+        $res = 0;
+        $criterionId = $criterion->getId();
+        foreach ($criterion->getNames() as $key => $value) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb
+            ->insert(_DB_PREFIX_ . 'product_comment_criterion_lang')
+            ->values(
+                [
+                    'id_product_comment_criterion' => '?',
+                    'id_lang' => '?',
+                    'name' => '?',
+                ]
+            )
+            ->setParameter(0, $criterionId)
+            ->setParameter(1, $key)
+            ->setParameter(2, $value)
+            ;
+            $res += $this->connection->executeUpdate($qb->getSQL(), $qb->getParameters(), $qb->getParameterTypes());
+        }
+
+        return $res;
+    }
+
+    private function updateCategories($criterion): int
+    {
+        $res = 0;
+        $criterionId = $criterion->getId();
+        foreach ($criterion->getCategories() as $id_category) {
+            $res += $this->connection->executeUpdate(
+                'INSERT INTO `' .
+                _DB_PREFIX_ . 'product_comment_criterion_category` (`id_product_comment_criterion`, `id_category`)
+                VALUES(' . $criterionId . ',' . $id_category . ')'
+            );
+        }
+
+        return $res;
+    }
+
+    private function updateProducts($criterion): int
+    {
+        $res = 0;
+        $criterionId = $criterion->getId();
+        foreach ($criterion->getProducts() as $id_product) {
```

Comment:
I maybe say something wrong but method name is `updateProducts()` however it does not update products. It creates new product_comment_criterion_product which indicate "this product comment is linked with this product".

So I see the creation but what about the deletion?

If I have comment A linked to products X, Y and Z.
I want to update comment A to link it to products X, Y and W. I dont want comment A to be linked to Z. This function cannot do it?

Reply:
It is related to [src/Entity/ProductCommentCriterion.php](https://github.com/PrestaShop/productcomments/pull/177/files#diff-e62c70a054534c7be80a40ecc106507cd61896d8ed136cacc6b7c21abdab439cR86). In that context, `updateProducts()`  means update `products` that linked to  `$criterion`.
For deletion, you can see function [deleteProducts](https://github.com/PrestaShop/productcomments/pull/177/files#diff-1d36351690ee35c0a5e1d5c5c323743183c86e0239242510597a76805a72bebbR105-R110) at 100 lines above, in the same file.
If you check [the diff](https://github.com/PrestaShop/productcomments/pull/177/files#diff-1d36351690ee35c0a5e1d5c5c323743183c86e0239242510597a76805a72bebbR105-R110) you will see I don't change the logic: delete all old linked products, then add new linked products.

Reply:
FYI, `updateProducts` and `deleteProducts` in  `ProductCommentCriterionRepository` is just a temporary solution when standard Doctrine `products` relation, `getProducts`, `addProduct`, `removeProduct` functions  are not implemented yet in `ProductCommentCriterion` Entity.